tastyworks/tastyworks_api/tw_api.py

```diff
@@ -0,0 +1,647 @@
+                   f'{"&" if journal_page_offset else "?"}'
+                   f'{"tag="+tag if tag else ""}',  # Used to search for an entry by tag
+        
+        'INST_FUTURES': '/instruments/futures',
+        'PRECISIONS': '/instruments/quantity-decimal-precisions'
+    }
+
+    url = root + crumbs.get(key)
+    return url
+
+
+async def api_request(request_type: str, url: str, token: str = None, json_data: dict = None):
+    """
+    Get all the data from the TW API from a specific API URL and request type
+
+    Args:
+        request_type (str): Type of request ('GET', 'POST')
+        url (str): url returned by APIUrl.METRICS.make_url()
+        token (str): The TW API token from an open session
+        json_data (dict): The JSON payload associated with the request
+    Returns:
+        The server response
+    """
+    print(f'##########\n'
+          f'Running: {inspect.stack()[1].function}\n'
+          f'##########')
+
+    if token is not None:
+        header = {'Authorization': token}
+    else:
+        header = {}
+
+    async with aiohttp.request(request_type, url, headers=header, json=json_data) as resp:
+        await is_error(resp)
+        if resp:
+            api_response = {
+                'content': 'OK' if (resp.status in [202, 204]) else await resp.json(),  # 202/204 not returning any data
+                'ok': resp.ok,
+                'status': resp.status,
+                'reason': resp.reason,
+                'url': str(resp.url),
+                'method': resp.method,
+                'content_length': resp.content_length,
+                'content_type': resp.content_type
+            }
+        else:
+            api_response = {}
+
+    return api_response
+
+
+"""
+#########################################
+############### SESSION #################
+#########################################
+"""
+
+
+async def session_start(username: str = None, password: str = None) -> Dict:
+    """
+    Start a TW API session.
+
+    Args:
+        username: TW account username
+        password: TW account password
+
+    Returns:
+        The API server response containing the session token as well as user/account data
+    """
+    if username is None and password is None:
+        body = {
+            'login': environ.get('TW_USER', ""),
+            'password': environ.get('TW_PASSWORD', "")
+        }
+    else:
+        body = {
+            'login': username,
+            'password': password
+        }
+
+    url = make_api_url('SESSION_START')
+    resp = await api_request('POST', url, json_data=body)
+    return resp
+
+
+async def session_validate(token: str) -> Dict:
+    """
+    Validate the status of a TW API session using the session token
+    """
+    url = make_api_url('SESSION_VALIDATE')
+    resp = await api_request('POST', url, token)
+    return resp
+
+
+async def get_streamer_info(token: str) -> Dict:
```

Comment:
Just noting this one.  No complaints, I'm just not sure how useful this is outside of `streamer.py`.